### PR TITLE
Fixing the profile loader

### PR DESCRIPTION
**Pull request by enkessler**
Fixed a profile parsing bug that was only affecting Windows.

## Summary

A bug in the profile loader was discovered while working on #1116. This PR fixes that bug.

## Details

Bug fix + tests. Rather than try to re-implement the parsing being done by Shellwords, I simply _temporarily_ adjusted the input data so that Shellwords could handle it as usual.
 
## Motivation and Context

The problem was discovered during #1116 but it was decided to handle it separately rather than making it a blocking problem on that PR.

## How Has This Been Tested?

New spec added for the profile loader.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] I've added tests for my code
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: spec/cucumber/cli/profile_loader_spec.rb [hunk 1]
```diff
@@ -42,6 +42,16 @@ def loader
         given_cucumber_yml_defined_as yml
         expect(loader.args_from('default')).to eq ['--format','pretty','features/sync_imap_mailbox.feature:16:22']
       end
+
+      it 'correctly parses a profile that uses tag expressions' do
```
Comment (brasmusson) [inline spec/cucumber/cli/profile_loader_spec.rb]: I would also add an example using `'` around the argument containing blanks, for instance `"--format pretty features --tags 'not @jruby'"` (which also avoids the path separator problem), to be very sure that both `"` and `'` can be used to define arguments with blanks in them.

Comment (enkessler) [inline spec/cucumber/cli/profile_loader_spec.rb]: Done.

Comment (enkessler) [inline spec/cucumber/cli/profile_loader_spec.rb]: I don't know why this PR isn't noticing the new commit that I made. The new test has been added to this branch but it's not showing up...

Comment (enkessler) [inline spec/cucumber/cli/profile_loader_spec.rb]: There it is. GitHub is slow today. Anything else @brasmusson?

Comment (brasmusson) [inline spec/cucumber/cli/profile_loader_spec.rb]: Now, it looks good to me.

**Comment by olleolleolle** [comment]
@enkessler _Maybe_ there is a quick way to green about these RuboCop warnings:

- `rubocop --auto-correct spec/cucumber/cli/profile_loader_spec.rb` would get rid of the `{ ... }` complaints
- and the one about `Exception` being too wide a rescue, it can be ` # rubocop:disable Lint/RescueException`

**Comment by enkessler** [comment]
Strange. The rubocop task doesn't have any issues when I run it locally.

**Comment by nodo** [comment]
@enkessler It's because the file is ignored by rubocop. You need to change `.rubocop_todo.yml` first.

**Comment by enkessler** [comment]
Ah, right. I read your comment too quickly.

**Comment by enkessler** [comment]
@olleolleolle I opted to just go ahead and change the exception type. Not catching _everything_ is probably good advice and `StandardError` is what is being caught in other areas of the code. I checked out the history and nothing indicates that it was a broader net intentionally. Maybe that's just how the contributor wrote code 8 years ago.

File: spec/cucumber/cli/profile_loader_spec.rb [hunk 2]
```diff
@@ -42,6 +42,25 @@ def loader
         given_cucumber_yml_defined_as yml
         expect(loader.args_from('default')).to eq ['--format','pretty','features/sync_imap_mailbox.feature:16:22']
       end
+
+      it 'correctly parses a profile that uses tag expressions (with double quotes)' do
```
Comment (mattwynne) [inline spec/cucumber/cli/profile_loader_spec.rb]: I'm wondering if this test name is misleading. It seems to me that the bug we've found is to do with parsing *any* switches given after the feature file paths, and that tag expressions just happen to be the example we've found of that general defective behaviour. Am I right?

File: spec/cucumber/cli/profile_loader_spec.rb [hunk 3]
```diff
@@ -42,6 +42,25 @@ def loader
...
+        given_cucumber_yml_defined_as('default' => '--format "pretty" features\sync_imap_mailbox.feature:16:22 --tags "not @jruby"')
+        if(Cucumber::WINDOWS)
+          expect(loader.args_from('default')).to eq ['--format','pretty','features\sync_imap_mailbox.feature:16:22','--tags','not @jruby']
+        else
+          expect(loader.args_from('default')).to eq ['--format','pretty','featuressync_imap_mailbox.feature:16:22','--tags','not @jruby']
+        end
+      end
+
+      it 'correctly parses a profile that uses tag expressions (with single quotes)' do
+        given_cucumber_yml_defined_as('default' => "--format 'pretty' features\\sync_imap_mailbox.feature:16:22 --tags 'not @jruby'")
```
Comment (mattwynne) [inline spec/cucumber/cli/profile_loader_spec.rb]: I don't really understand what's going on with the backslashes in these tests! 😁 

This one uses two, but the one on line 47 uses one. And the expected parsed output on non-windows OSes has no slashes in the path?!

I realise this is code we've inherited from our ancestors, but now that you've looked at it, are you able to explain what it means @enkessler?

Comment (brasmusson) [inline spec/cucumber/cli/profile_loader_spec.rb]: @mattwynne From the irb:
```
irb(main):004:0> puts 'dir\file'
dir\file
=> nil
irb(main):005:0> puts "dir\file"
dir
   ile
=> nil
irb(main):006:0> puts "dir\\file"
dir\file
=> nil
```
Within `""` the `\` is interpreted as "escape" so if you literally mean `\` you have to use `\\`, which when the `""` expression is "evaluted" is "converted" to `\`. (`\f` in the `""` expression above means ["formfead"](https://stackoverflow.com/a/31791519))

Comment (mattwynne) [inline spec/cucumber/cli/profile_loader_spec.rb]: Thanks @brassmusson, that makes sense.

I'm still puzzled why the expected output on non-windows has no slashes in the path at all though. What's going on there?!

File: spec/cucumber/cli/profile_loader_spec.rb (hunk 2 above)
Comment (enkessler) [inline spec/cucumber/cli/profile_loader_spec.rb]: I think you're right.

File: spec/cucumber/cli/profile_loader_spec.rb (hunk 3 above)
Comment (enkessler) [inline spec/cucumber/cli/profile_loader_spec.rb]: I don't know why the output is different either but I just copy/paste/tweaked for the new test.

As for why a certain number of backslashes are needed...well...they are basically [magic](https://xkcd.com/1638/). ;)

**Pull request merged**